File: ios/Omniverse/Screens/SearchScreen.swift
import SwiftUI

/// Faithful port of `search_screen.dart`.
/// Autofocus field, 320ms debounce, multi-source search, grid of poster cards,
/// navigation to MediaDetailScreen via NavigationStack.
struct SearchScreen: View {
    @Environment(AppState.self) private var state
    @Environment(\.dismiss) private var dismiss

    @State private var query = ""
    @State private var currentQuery = ""
    @State private var loading = false
    @State private var error: String?
    @State private var results: [MediaItem] = []
    @State private var requestId = 0
    @State private var debounceTask: Task<Void, Never>?
    @State private var path = NavigationPath()
    @FocusState private var fieldFocused: Bool

    var body: some View {
        NavigationStack(path: $path) {
            VStack(spacing: 0) {
                searchBar
                if loading {
                    ProgressView()
                        .tint(LiquidColors.cyan)
                        .padding(24)
                }
                if !loading, let error, results.isEmpty {
                    Text(error)
                        .font(.system(size: 17, weight: .semibold))
                        .foregroundStyle(.white.opacity(0.78))
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 24)
                        .padding(.top, 28)
                        .padding(.bottom, 12)
                }
                if results.isEmpty {
                    emptyHints
                } else {
                    resultsGrid
                }
                Spacer(minLength: 0)
            }
            .navigationDestination(for: MediaItem.self) { MediaDetailScreen(item: $0) }
            .toolbar(.hidden, for: .navigationBar)
        }
        .liquidScaffold()
        .onAppear { fieldFocused = true }
    }

    // MARK: - Search bar

    private var searchBar: some View {
        HStack(spacing: 8) {
            Button { dismiss() } label: {
                Image(systemName: "xmark").font(.system(size: 18, weight: .semibold)).foregroundStyle(.white)
                    .frame(width: 40, height: 40)
            }
            .buttonStyle(.plain)
            GlassCapsule(padding: EdgeInsets(top: 4, leading: 14, bottom: 4, trailing: 8)) {
                HStack(spacing: 10) {
                    Image(systemName: "magnifyingglass").font(.system(size: 18)).foregroundStyle(.white.opacity(0.7))
                    TextField("", text: $query, prompt: Text("Search movies and TV shows").foregroundColor(.white.opacity(0.54)))
                        .focused($fieldFocused)
                        .textInputAutocapitalization(.never)
                        .autocorrectionDisabled()
                        .submitLabel(.search)
                        .foregroundStyle(.white)
                        .font(.system(size: 17, weight: .semibold))
                        .tint(LiquidColors.cyan)
                        .onChange(of: query) { _, value in onChanged(value) }
                        .onSubmit { run(query.trimmed) }
                    if !query.isEmpty {
                        Button {
                            query = ""
                            onChanged("")
                        } label: {
                            Image(systemName: "xmark.circle.fill").font(.system(size: 18)).foregroundStyle(.white.opacity(0.7))
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
        }
        .padding(.horizontal, 12)
        .padding(.top, 10)
        .padding(.bottom, 6)
    }

    // MARK: - Results grid

    private var resultsGrid: some View {
        GeometryReader { geo in
            let count = columns(for: geo.size.width)
            ScrollView {
                LazyVGrid(columns: Array(repeating: GridItem(.flexible(), spacing: 14), count: count), spacing: 18) {
                    ForEach(results) { item in
                        Button { path.append(item) } label: { GridPoster(item: item) }
                            .buttonStyle(.plain)
                    }
                }
                .padding(EdgeInsets(top: 8, leading: 16, bottom: 32, trailing: 16))
            }
            .scrollIndicators(.hidden)
        }
    }

    private func columns(for width: CGFloat) -> Int {
        if width >= 1200 { return 6 }
        if width >= 900 { return 5 }
        if width >= 600 { return 4 }
        return 3
    }

    // MARK: - Empty hints

    @ViewBuilder private var emptyHints: some View {
        if currentQuery.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                Text("Search Anything in Omniverse")
                    .font(.system(size: 19, weight: .black))
                    .foregroundStyle(.white)
                Text("Type a movie, TV show, or anime title. We search Omniverse for matches and open the same detail screen as the home rows — sources included.")
                    .font(.system(size: 15))
                    .foregroundStyle(.white.opacity(0.7))
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(EdgeInsets(top: 8, leading: 28, bottom: 0, trailing: 28))
        }
    }

    // MARK: - Debounce + run

    private func onChanged(_ value: String) {
        debounceTask?.cancel()
        let trimmed = value.trimmed
        if trimmed.isEmpty {
            currentQuery = ""
            results = []
            loading = false
            error = nil
            return
        }
        debounceTask = Task {
            try? await Task.sleep(nanoseconds: 320_000_000)
            if Task.isCancelled { return }
            await runSearch(trimmed)
        }
    }

    private func run(_ q: String) { Task { await runSearch(q) } }

    @MainActor
    private func runSearch(_ q: String) async {
        let trimmed = q.trimmed
        guard !trimmed.isEmpty else { return }
        requestId += 1
        let id = requestId
        currentQuery = trimmed
        loading = true
        error = nil

        guard state.credentials.hasTmdb else {
            if id != requestId { return }
            loading = false
            error = "Add your TMDB API key in Settings to enable search."
            results = []
            return
        }
        let items = await state.searchMedia(trimmed)
        if id != requestId { return }
        loading = false
        results = items
        error = items.isEmpty ? "No matches for \"\(trimmed)\"." : nil
    }
}

/// Grid poster (2:3) sized to fill its grid cell.
private struct GridPoster: View {
    let item: MediaItem
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            PosterImage(url: item.posterUrl ?? item.backdropUrl)
                .aspectRatio(2/3, contentMode: .fill)
                .frame(maxWidth: .infinity)
                .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: 12, style: .continuous).strokeBorder(Color.white.opacity(0.12), lineWidth: 1))
                .shadow(color: .black.opacity(0.36), radius: 14, y: 8)
            Text(item.title).font(.system(size: 13, weight: .bold)).foregroundStyle(.white).lineLimit(1)
        }
    }
}

File: android/app/src/main/java/com/finix/omniverse/ui/SearchScreen.kt
package com.finix.omniverse.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.finix.omniverse.AppGraph
import com.finix.omniverse.MediaItem
import com.finix.omniverse.ui.theme.LiquidColors
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged

@OptIn(kotlinx.coroutines.FlowPreview::class)
@Composable
fun SearchScreen(nav: NavController) {
    val state = AppGraph.appState
    var query by remember { mutableStateOf("") }
    var currentQuery by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var results by remember { mutableStateOf<List<MediaItem>>(emptyList()) }
    var requestId by remember { mutableIntStateOf(0) }

    // 320ms debounce on query
    LaunchedEffect(Unit) {
        snapshotFlow { query }
            .debounce(320)
            .distinctUntilChanged()
            .collect { value ->
                val trimmed = value.trim()
                if (trimmed.isEmpty()) { currentQuery = ""; results = emptyList(); loading = false; error = null; return@collect }
                requestId++
                val id = requestId
                currentQuery = trimmed
                loading = true; error = null
                if (!state.credentials.hasTmdb) {
                    loading = false; error = "Add your TMDB API key in Settings to enable search."; results = emptyList()
                    return@collect
                }
                val items = state.searchMedia(trimmed)
                if (id != requestId) return@collect
                loading = false; results = items
                error = if (items.isEmpty()) "No matches for \"$trimmed\"." else null
            }
    }

    Column(Modifier.fillMaxSize().statusBarsPadding()) {
        Row(Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 10.dp), verticalAlignment = Alignment.CenterVertically) {
            OutlinedTextField(
                value = query, onValueChange = { query = it },
                modifier = Modifier.fillMaxWidth(),
                placeholder = { Text("Search movies and TV shows", color = Color.White.copy(alpha = 0.54f)) },
                leadingIcon = { Icon(Icons.Filled.Search, null, tint = Color.White.copy(alpha = 0.7f)) },
                singleLine = true,
                shape = RoundedCornerShape(50),
                colors = OutlinedTextFieldDefaults.colors(
                    focusedTextColor = Color.White, unfocusedTextColor = Color.White,
                    focusedBorderColor = LiquidColors.Cyan, unfocusedBorderColor = Color.White.copy(alpha = 0.2f),
                    cursorColor = LiquidColors.Cyan,
                ),
            )
        }
        if (loading) Box(Modifier.fillMaxWidth().padding(24.dp), Alignment.Center) { CircularProgressIndicator(color = LiquidColors.Cyan) }
        if (!loading && error != null && results.isEmpty()) {
            Text(error!!, color = Color.White.copy(alpha = 0.78f), fontSize = 17.sp, fontWeight = FontWeight.SemiBold,
                modifier = Modifier.fillMaxWidth().padding(horizontal = 24.dp, vertical = 20.dp))
        }
        if (results.isEmpty() && currentQuery.isEmpty()) {
            Column(Modifier.padding(horizontal = 28.dp, vertical = 8.dp)) {
                Text("Search Anything in Omniverse", color = Color.White, fontSize = 19.sp, fontWeight = FontWeight.Black)
                Text("Type a movie, TV show, or anime title. We search Omniverse for matches and open the same detail screen as the home rows — sources included.",
                    color = Color.White.copy(alpha = 0.7f), fontSize = 15.sp)
            }
        } else {
            androidx.compose.foundation.layout.BoxWithConstraints(Modifier.fillMaxSize()) {
                val cols = when { maxWidth >= 1200.dp -> 6; maxWidth >= 900.dp -> 5; maxWidth >= 600.dp -> 4; else -> 3 }
                LazyVerticalGrid(
                    columns = GridCells.Fixed(cols),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(start = 16.dp, end = 16.dp, top = 8.dp, bottom = 32.dp),
                    horizontalArrangement = Arrangement.spacedBy(14.dp),
                    verticalArrangement = Arrangement.spacedBy(18.dp),
                ) {
                    items(results, key = { it.id }) { item ->
                        Column {
                            Box(
                                Modifier.fillMaxWidth().aspectRatio(2f / 3f)
                                    .tvFocusable(onClick = { RouteArgs.detailItem = item; nav.navigate("detail") })
                                    .border(1.dp, Color.White.copy(alpha = 0.12f), RoundedCornerShape(12.dp))
                            ) { PosterImage(item.posterUrl ?: item.backdropUrl, Modifier.fillMaxSize().clip(RoundedCornerShape(12.dp))) }
                            Text(item.title, color = Color.White, fontSize = 13.sp, fontWeight = FontWeight.Bold, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.padding(top = 8.dp))
                        }
                    }
                }
            }
        }
    }
}
